src/main/scala/is/hail/io/bgen/BgenRDDPartitions.scala

```diff
@@ -208,5 +236,312 @@ object BgenRDDPartitions extends Logging {
+            nRow := reader.invoke[Int]("readInt"),
+            (nRow.cne(settings.nSamples)).mux(
+              Code._fatal(
+                const("row nSamples is not equal to header nSamples ")
+                  .concat(nRow.toS)
+                  .concat(", ")
+                  .concat(settings.nSamples.toString)),
+              Code._empty),
+            nAlleles2 := reader.invoke[Int]("readShort"),
+            (nAlleles.cne(nAlleles2)).mux(
+              Code._fatal(
+                const("""Value for `nAlleles' in genotype probability data storage is
+                        |not equal to value in variant identifying data. Expected""".stripMargin)
+                  .concat(nAlleles.toS)
+                  .concat(" but found ")
+                  .concat(nAlleles2.toS)
+                  .concat(" at ")
+                  .concat(contig)
+                  .concat(":")
+                  .concat(position.toS)
+                  .concat(".")),
+              Code._empty),
+            minPloidy := reader.invoke[Int]("read"),
+            maxPloidy := reader.invoke[Int]("read"),
+            (minPloidy.cne(2) || maxPloidy.cne(2)).mux(
+              Code._fatal(
+                const("Hail only supports diploid genotypes. Found min ploidy equals `")
+                  .concat(minPloidy.toS)
+                  .concat("' and max ploidy equals `")
+                  .concat(maxPloidy.toS)
+                  .concat("'.")),
+              Code._empty),
+            i := 0,
+            Code.whileLoop(i < settings.nSamples,
+              ploidy := reader.invoke[Int]("read"),
+              ((ploidy & 0x3f).cne(2)).mux(
+                Code._fatal(
+                  const("Ploidy value must equal to 2. Found ")
+                    .concat(ploidy.toS)
+                    .concat(".")),
+                Code._empty),
+              i += 1
+            ),
+            phase := reader.invoke[Int]("read"),
+            (phase.cne(0) && phase.cne(1)).mux(
+              Code._fatal(
+                const("Value for phase must be 0 or 1. Found ")
+                  .concat(phase.toS)
+                  .concat(".")),
+              Code._empty),
+            phase.ceq(1).mux(
+              Code._fatal("Hail does not support phased genotypes."),
+              Code._empty),
+            nBitsPerProb := reader.invoke[Int]("read"),
+            (nBitsPerProb < 1 || nBitsPerProb > 32).mux(
+              Code._fatal(
+                const("Value for nBits must be between 1 and 32 inclusive. Found ")
+                  .concat(nBitsPerProb.toS)
+                  .concat(".")),
+              Code._empty),
+            (nBitsPerProb.cne(8)).mux(
+              Code._fatal(
+                const("Only 8-bit probabilities supported, found ")
+                  .concat(nBitsPerProb.toS)
+                  .concat(".")),
+              Code._empty),
+            nGenotypes := Code.invokeScalaObject[Int, Int](
```

Comment:
The code is styled to stay generic until we get to probability loop... e.g. at this point we know that `nAllelles` is 2 and `nGenotypes` is 3 and `nBitsPerProb` is 8, but don't take advantage. Is this style a consequence of having previously supporting the more general case, or meant to make it easier to generalize later, or meant to be conceptually clearer. My guess is all of the above and I'm fine keeping it since it's ahead of the expensive loop.

Reply:
This is the way the code was written before I transplanted it here. I'll hardcode all the values.